flow/layers/physical_shape_layer.cc

```diff
@@ -5,155 +5,147 @@
+    // opacity.
+    if (context->is_opaque) {
+      set_paint_bounds(SkRect::MakeEmpty());
     }
+    return;
   }
 
-  UpdateSceneChildren(context);
+  // Add some margin to the paint bounds to leave space for the shadow.
+  // We fill this whole region and clip children to it so we don't need to
+  // join the child paint bounds.
+  // The offset is calculated as follows:
+  //                   .---                           (kLightRadius)
+  //                -------/                          (light)
+  //                   |  /
+  //                   | /
+  //                   |/
+  //                   |O
+  //                  /|                              (kLightHeight)
+  //                 / |
+  //                /  |
+  //               /   |
+  //              /    |
+  //             -------------                        (layer)
+  //            /|     |
+  //           / |     |                              (elevation)
+  //        A /  |     |B
+  // ------------------------------------------------ (canvas)
+  //          ---                                     (extent of shadow)
+  //
+  // E = lt        }           t = (r + w/2)/h
+  //                } =>
+  // r + w/2 = ht  }           E = (l/h)(r + w/2)
+  //
+  // Where: E = extent of shadow
+  //        l = elevation of layer
+  //        r = radius of the light source
+  //        w = width of the layer
+  //        h = light height
+  //        t = tangent of AOB, i.e., multiplier for elevation to extent
+  SkRect bounds(paint_bounds());
+  // tangent for x
+  double tx = (kLightRadius * context->frame_device_pixel_ratio +
+               bounds.width() * 0.5) /
+              kLightHeight;
+  // tangent for y
+  double ty = (kLightRadius * context->frame_device_pixel_ratio +
+               bounds.height() * 0.5) /
+              kLightHeight;
+  bounds.outset(elevation() * tx, elevation() * ty);
+  set_paint_bounds(bounds);
 }
 
-#endif  // defined(OS_FUCHSIA)
-
 void PhysicalShapeLayer::Paint(PaintContext& context) const {
   TRACE_EVENT0("flutter", "PhysicalShapeLayer::Paint");
   FML_DCHECK(needs_painting());
 
-  if (elevation_ != 0) {
-    DrawShadow(context.leaf_nodes_canvas, path_, shadow_color_, elevation_,
-               SkColorGetA(color_) != 0xff, device_pixel_ratio_);
+  // The compositor will paint this layer (which is a solid color) via the
+  // color on |SceneUpdateContext::Frame|.
+  //
+  // The child layers will be painted into the texture used by the Frame, so
+  // painting them here would actually cause them to be painted on the display
+  // twice -- once into the current canvas (which may be inside of another
+  // Frame) and once into the Frame's canvas (which is then stacked on top of
+  // the current canvas).
+  if (PhysicalShapeLayerBase::should_system_composite()) {
+    FML_DCHECK(needs_system_composite());
+
+    PhysicalShapeLayerBase::Paint(context);
+    return;
+  }
+
+  if (elevation() != 0) {
```

Comment:
can we add trace events for this?

Reply:
There is a trace event in TRACE_EVENT0("flutter", "FuchsiaSystemCompositedContainerLayer::Paint"); for this

Reply:
FuchsiaSystemCompositedContainerLayer has a trace event for this